### PR TITLE
Replace Hazel by stack_snapshot

### PR DESCRIPTION
**Pull request by aherrmann-da**
- Replace the now deprecated Hazel by `rules_haskell`'s `stack_snapshot` which uses the `stack` to fetch Hackage dependencies and the `Cabal` library to build them.
- Similar to how JVM dependencies are defined in `bazel-java-deps.bzl`, Haskell dependencies are now defined in `bazel-haskell-deps.bzl`.
- Avoid Bazel name mangling and generate `libz.so` and `libbz2.so` for external library dependencies. `Cabal` requires library dependencies to match the expected name.
- Update `rules_haskell` for `stack_snapshot` support.
- Add patches for additional requred `stack_snapshot` features. These patches can be removed once the required features have landed upstream.
- Most hackage dependencies are defined in a custom stack snapshot in `stack-snapshot.yaml`.
- Where patches or custom builds are required, "vendored packages" are defined in the `bazel-haskell-deps` file. E.g. `ghcide`.
- Binaries are not generated by `stack_snapshot`, so these are defined in `bazel-haskell-deps.bzl`. Their library dependencies are supplied using the `stack_snapshot`.
- Updates the documentation from Hazel to `stack_snapshot`.
- Removes Hazel code.

Easiest reviewed commit by commit.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by aherrmann-da** [comment]
Windows CI fails with
```
HttpExceptionRequest Request {
  host                 = "digitalassetsdk.bintray.com"
  port                 = 443
  secure               = True
  requestHeaders       = [("User-Agent","Haskell pantry package")]
  path                 = "/ghc-lib/ghc-lib-parser-8.8.1.20190912.tar.gz"
  queryString          = ""
  method               = "GET"
  proxy                = Nothing
  rawBody              = False
  redirectCount        = 10
  responseTimeout      = ResponseTimeoutDefault
  requestVersion       = HTTP/1.1
}
 (InternalException (HandshakeFailed (Error_Protocol ("certificate rejected: [SelfSigned]",True,CertificateUnknown))))
```
Windows CI on this PR used to not have this certificate issue, e.g. [this run](https://dev.azure.com/digitalasset/daml/_build/results?buildId=16571) reached the stage of building `@stackage` packages, meaning that fetching `ghc-lib(-parser)` did succeed.

File: WORKSPACE [hunk 1]
```diff
@@ -32,6 +30,471 @@ load("@rules_haskell//tools:repositories.bzl", "rules_haskell_worker_dependencie
...
+    srcs = glob(["**"]),
+    visibility = ["//visibility:public"],
+)
+    """,
+    sha256 = "9094d19ed0db980a34f1ffd58e64c7df9b4ecb3beed22fd9b9739044a8d45f77",
+    strip_prefix = "happy-1.19.11",
+    urls = ["http://hackage.haskell.org/package/happy-1.19.11/happy-1.19.11.tar.gz"],
+)
+
+http_archive(
```
Comment (cocreature) [inline WORKSPACE]: I guess there is no easier way to get a package from Hackage that isn’t in Stackage? This seems rather verbose and annoying to maintain.

File: bazel_tools/haskell-cabal-collisions.patch [hunk 2]
```diff
@@ -0,0 +1,290 @@
+From 17a2c0e7813236fdcc8107544922d44fcabe924c Mon Sep 17 00:00:00 2001
```
Comment (cocreature) [inline bazel_tools/haskell-cabal-collisions.patch]: I guess we can upstream this in the future?

File: WORKSPACE (hunk 1 above)
Comment (cocreature) [reply WORKSPACE]: Actually I’m not sure why we can’t modify the `stack-snapshot.yaml` file for these?

File: bazel_tools/haskell-cabal-cc-wrapper.patch [hunk 3]
```diff
@@ -0,0 +1,49 @@
+diff --git a/haskell/cabal.bzl b/haskell/cabal.bzl
```
Comment (cocreature) [inline bazel_tools/haskell-cabal-cc-wrapper.patch]: I suppose that’s upstreameable as well?

File: bazel_tools/haskell-stack-snapshot-deps.patch [hunk 4]
```diff
@@ -0,0 +1,144 @@
+From 14272ac578adbbfd1737f9f191951c2a49856120 Mon Sep 17 00:00:00 2001
```
Comment (cocreature) [inline bazel_tools/haskell-stack-snapshot-deps.patch]: Also looks ustreamable?

File: WORKSPACE (hunk 1 above)
Comment (aherrmann-da) [reply WORKSPACE]: The reason is that this is an executable target, not a library target. That's what I meant by
> Binaries are not generated by stack_snapshot, so these are defined in WORKSPACE. Their library dependencies are supplied using the stack_snapshot.

I.e. the library `@stackage//:hpp` is generated by `stack_snapshot`, but the binary is defined separately.

File: bazel_tools/haskell-cabal-collisions.patch (hunk 2 above)
Comment (aherrmann-da) [reply bazel_tools/haskell-cabal-collisions.patch]: Yes, there's a PR for it already, it's linked in `deps.bzl` https://github.com/digital-asset/daml/pull/2743/files#diff-0d048ba7b51a63feefd2338106bd6d99R53

File: bazel_tools/haskell-cabal-cc-wrapper.patch (hunk 3 above)
Comment (aherrmann-da) [reply bazel_tools/haskell-cabal-cc-wrapper.patch]: Yes, I haven't open a PR, yet. But, I'll mark it and others as upstreamable.

File: bazel_tools/haskell-stack-snapshot-deps.patch (hunk 4 above)
Comment (aherrmann-da) [reply bazel_tools/haskell-stack-snapshot-deps.patch]: Yes, there's also a PR for this one. It's linked here https://github.com/digital-asset/daml/pull/2743/files#diff-0d048ba7b51a63feefd2338106bd6d99R51

File: WORKSPACE (hunk 1 above)
Comment (cocreature) [reply WORKSPACE]: Ah I see makes sense, thanks for the explanation! That’s what I get for not reading the docs closely :slightly_smiling_face:

**Pull request merged**